cdap-app-fabric/src/main/java/co/cask/cdap/internal/provision/ProvisioningService.java

```diff
@@ -321,6 +332,23 @@ public Runnable provision(ProvisionRequest provisionRequest, DatasetContext data
       return () -> { };
     }
 
+    // get plugin requirement information and check for capability to run on the provisioner
+    Set<PluginRequirement> requirements = GSON.fromJson(args.get(ProgramOptionConstants.PLUGIN_REQUIREMENTS),
+                                                        PLUGIN_REQUIREMENT_SET_TYPE);
+    if (requirements != null) {
+      Set<PluginRequirement> unfulfilledRequirements = getUnfulfilledRequirements(provisioner.getCapabilities(),
+                                                                                  requirements);
+      if (!unfulfilledRequirements.isEmpty()) {
+        runWithProgramLogging(programRunId, args, () ->
+          LOG.error(String.format("'%s' cannot be run using profile '%s' because the profile does not met all " +
+                                    "plugin requirements. Following requirements were not meet by the listed " +
+                                    "plugins: '%s'", programRunId.getProgram(), name,
+                                  groupByRequirement(unfulfilledRequirements))));
```

Comment:
now that my other PR is merged, need to call `programStateWriter.error()` before sending the deprovisioned message.